core/trino-main/src/main/java/io/trino/operator/scalar/ArraySubscriptOperator.java

```diff
@@ -64,94 +80,46 @@ protected ArraySubscriptOperator()
+        Class<?> expectedReturnType = methodType(elementType.getJavaType()).wrap().returnType();
+        methodHandle = explicitCastArguments(methodHandle, methodHandle.type().changeReturnType(expectedReturnType));
+        methodHandle = guardWithTest(
+                IS_POSITION_NULL,
+                empty(methodHandle.type()),
+                methodHandle);
+        methodHandle = collectArguments(methodHandle, 1, GET_POSITION);
+        methodHandle = permuteArguments(methodHandle, methodHandle.type().dropParameterTypes(1, 2), 0, 0, 1);
 
-        MethodHandle methodHandle;
-        if (elementType.getJavaType() == boolean.class) {
-            methodHandle = METHOD_HANDLE_BOOLEAN;
-        }
-        else if (elementType.getJavaType() == long.class) {
-            methodHandle = METHOD_HANDLE_LONG;
-        }
-        else if (elementType.getJavaType() == double.class) {
-            methodHandle = METHOD_HANDLE_DOUBLE;
-        }
-        else if (elementType.getJavaType() == Slice.class) {
-            methodHandle = METHOD_HANDLE_SLICE;
-        }
-        else {
-            methodHandle = METHOD_HANDLE_OBJECT.asType(
-                    METHOD_HANDLE_OBJECT.type().changeReturnType(elementType.getJavaType()));
-        }
-        methodHandle = methodHandle.bindTo(elementType);
-        requireNonNull(methodHandle, "methodHandle is null");
         return new ChoicesSpecializedSqlScalarFunction(
                 boundSignature,
                 NULLABLE_RETURN,
                 ImmutableList.of(NEVER_NULL, NEVER_NULL),
                 methodHandle);
     }
 
-    @UsedByGeneratedCode
-    public static Long longSubscript(Type elementType, Block array, long index)
-    {
-        checkIndex(array, index);
-        int position = toIntExact(index - 1);
-        if (array.isNull(position)) {
-            return null;
-        }
-
-        return elementType.getLong(array, position);
-    }
-
-    @UsedByGeneratedCode
-    public static Boolean booleanSubscript(Type elementType, Block array, long index)
+    private static int getPosition(Block array, long index)
     {
-        checkIndex(array, index);
-        int position = toIntExact(index - 1);
-        if (array.isNull(position)) {
-            return null;
-        }
-
-        return elementType.getBoolean(array, position);
-    }
-
-    @UsedByGeneratedCode
-    public static Double doubleSubscript(Type elementType, Block array, long index)
-    {
-        checkIndex(array, index);
-        int position = toIntExact(index - 1);
-        if (array.isNull(position)) {
-            return null;
-        }
-
-        return elementType.getDouble(array, position);
-    }
-
-    @UsedByGeneratedCode
-    public static Slice sliceSubscript(Type elementType, Block array, long index)
-    {
-        checkIndex(array, index);
-        int position = toIntExact(index - 1);
-        if (array.isNull(position)) {
-            return null;
+        checkArrayIndex(index);
```

Comment:
You can inline this now that there is only one usage. Also, consider converting to `checkCondition`